bundles/org.openhab.binding.evcc/src/main/java/org/openhab/binding/evcc/internal/EvccAPI.java

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) 2010-2022 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.evcc.internal;
+
+import static org.openhab.binding.evcc.internal.EvccBindingConstants.EVCC_REST_API;
+import static org.openhab.binding.evcc.internal.EvccBindingConstants.LONG_CONNECTION_TIMEOUT_MILLISEC;
+
+import java.io.IOException;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.evcc.internal.dto.Result;
+import org.openhab.binding.evcc.internal.dto.Status;
+import org.openhab.core.io.net.http.HttpUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link EvccAPI} is responsible for API calls to evcc.
+ * 
+ * @author Florian Hotze - Initial contribution
+ */
+@NonNullByDefault
+public class EvccAPI {
+    private static final Logger logger = LoggerFactory.getLogger(EvccAPI.class);
+    private static final Gson gson = new Gson();
+    private String host = "";
+
+    public EvccAPI(String host) {
+        this.host = host;
+    }
+
+    /**
+     * Make a HTTP request.
+     * 
+     * @param description request description for logger
+     * @param url full request URL
+     * @param method reguest method, e.g. GET, POST
+     * @return the response body or null if request failed
+     */
+    private @Nullable String httpRequest(@Nullable String description, String url, String method) {
+        String response = "";
+        try {
+            response = HttpUtil.executeUrl(method, url, LONG_CONNECTION_TIMEOUT_MILLISEC);
+            logger.trace("{} - {}, {} - {}", description, url, method, response);
+            return response;
+        } catch (IOException e) {
+            logger.debug("IO Exception - {} - {}, {} - {}", description, url, method, e.toString());
+            return null;
```

Comment:
Which exception should I throw then?

Reply:
I was thinking some custom exception where you could include URL and method, but actually `IOException` could be sufficient for this case, so you could simply add "throws IOException" to the method signature and remove the local handling here. Then, in the handler, `handleCommand()`, you could catch it and handle it by logging it. You would need to adapt `getResult()` also for unified error handling, so either introduce custom exception which could be thrown from here, or handle both the Json exception and IOException in the handler.

Reply:
Hmm, after seeing last commits: I'm sorry if I got you started on this error handling refactoring without fully explaining my thoughts. So please just take the following as a description of my intentions here, and you can decide how much or little you want to do.

After defining a custom exception, for example:

```
@NonNullByDefault
public class EvccApiException extends Exception {

    private static final long serialVersionUID = 0L; // uniquely generated UID

    public EvccApiException(String message) {
        super(message);
    }

    public EvccApiException(String message, Throwable cause) {
        super(message, cause);
    }
}
```

this could then be thrown in case of failure:

```
private String httpRequest(String url, String method) throws EvccApiException {
    try {
        String response = HttpUtil.executeUrl(method, url, LONG_CONNECTION_TIMEOUT_MILLISEC);
        logger.trace("{}, {} - {}", url, method, response);
        return response;
    } catch (IOException e) {
        throw new EvccApiException("HTTP request failed for URL " + url, e);
    }
}
```

Then, in `getResult`:

```
public Result getResult() throws EvccApiException {
    final String response = httpRequest(this.host + EVCC_REST_API + "state", "GET");
    try {
        Status status = gson.fromJson(response, Status.class);
        if (status == null) {
            throw new EvccApiException("Status is null");
        }
        return status.getResult();
    } catch (JsonSyntaxException e) {
        throw new EvccApiException("Error parsing response: " + response, e);
    }
}
```

And the set methods:

```
public String setMode(int loadpoint, String mode) throws EvccApiException {
    return httpRequest(this.host + EVCC_REST_API + "loadpoints/" + loadpoint + "/mode/" + mode, "POST");
}
```

So now nothing is nullable and we only have one exception to deal with centrally - we'll either get a valid result or an exception thrown. So all we need to do in handler:

```
try {
    switch (channelIdWithoutGroup) {
        case CHANNEL_LOADPOINT_MODE:
            evccAPI.setMode(loadpoint, command.toString());
            break;
        case CHANNEL_LOADPOINT_MIN_SOC:
            evccAPI.setMinSoC(loadpoint, Integer.parseInt(command.toString().replaceAll(" %", "")));
            break;
} catch (EvccApiException e) {
    Throwable cause = e.getCause();
    if (cause == null) {
        logger.debug("Failed to update channel {}: {}", channelUID, e.getMessage());
    } else {
        logger.debug("Failed to update channel {}: {} -> {}", channelUID, e.getMessage(), cause.getMessage());
    }
}
```
(switch could even be extracted into separate private method for simplicity)

Reply:
That looks good, I will have a look into this.

```java
private static final long serialVersionUID = 0L; // uniquely generated UID
```

Do I need and if yes, how would I generate that UID?

Reply:
> (switch could even be extracted into separate private method for simplicity)

On one hand, yes, but on the other hand (at least for me) more methods make code more complicated to understand/read as you have to scroll more.

I will leave the switch inside of `handleCommand`

Reply:
>> private static final long serialVersionUID = 0L; // uniquely generated UID

> Do I need and if yes, how would I generate that UID?

Yes, you need to - your IDE can most likely generate it for you. At least I know IntelliJ and Eclipse can do it, so if you're using one of them, it would probably be easiest to google how to do it with your IDE and the specific version of it.

>> (switch could even be extracted into separate private method for simplicity)

> On one hand, yes, but on the other hand (at least for me) more methods make code more complicated to
> understand/read as you have to scroll more.

Well, I have to disagree here. :) By having smaller methods they will usually be simpler to read and understand as they would do fewer things, thus complexity would be smaller.

In the example here, having a private method for handling commands calling the set methods, let's call it `handleLoadpointCommand` (I don't know if it makes sense, so just for the example), it would be easy to see how errors are handled and that refresh() is called after calling the API set methods. Currently you will have to scroll down to understand this flow.

That said, I'm fine with your implementation, it was just a side note with an additional refactoring proposal. :)

Reply:
> Yes, you need to - your IDE can most likely generate it for you. At least I know IntelliJ and Eclipse can do it, so if you're using one of them, it would probably be easiest to google how to do it with your IDE and the specific version of it.

I found no way to do that in VS Code.
Maybe you could generate that UID in your IDE and post it here?

Reply:
I installed Ecplise, got the serialVersionUID generated.

See 7ade98f